File: apps/juris-mobile/lib/models/case_type_registry.dart
const String caseTypeRegistryId = 'genesis-juris-case-types';
const String caseTypeVersion = '1.0.0';

enum CaseTypeId {
  generalAdvisory('general_advisory'),
  litigationStrategy('litigation_strategy'),
  contractReview('contract_review'),
  taxPlanning('tax_planning'),
  compliance('compliance'),
  taxCompliance('tax_compliance'),
  erpIncident('erp_incident'),
  investigation('investigation'),
  trainingSimulation('training_simulation');

  const CaseTypeId(this.wireName);

  final String wireName;

  static CaseTypeId parse(String value) {
    return values.firstWhere(
      (CaseTypeId item) => item.wireName == value,
      orElse: () => throw const FormatException('Unknown case type.'),
    );
  }
}

enum StudioCaseViewId {
  issueMap('issue_map'),
  evidenceMap('evidence_map'),
  decisionTable('decision_table'),
  taskPlan('task_plan'),
  timeline('timeline'),
  economics('economics'),
  simulation('simulation');

  const StudioCaseViewId(this.wireName);

  final String wireName;
}

final class CaseTypeReference {
  const CaseTypeReference(this.id);

  final CaseTypeId id;

  String get registry => caseTypeRegistryId;
  String get version => caseTypeVersion;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'registry': registry,
        'id': id.wireName,
        'version': version,
      };

  factory CaseTypeReference.fromJson(Object? source) {
    if (source is! Map<String, dynamic> ||
        source['registry'] != caseTypeRegistryId ||
        source['id'] is! String ||
        source['version'] != caseTypeVersion) {
      throw const FormatException('Unsupported case-type package.');
    }
    return CaseTypeReference(CaseTypeId.parse(source['id'] as String));
  }
}

final class CaseTypeDefinition {
  const CaseTypeDefinition({
    required this.id,
    required this.workflowMode,
    required this.views,
    required this.requiredReview,
    required this.labelEn,
    required this.labelRu,
    required this.summaryEn,
    required this.summaryRu,
    required this.outcomeEn,
    required this.outcomeRu,
  });

  final CaseTypeId id;
  final String workflowMode;
  final List<StudioCaseViewId> views;
  final String requiredReview;
  final String labelEn;
  final String labelRu;
  final String summaryEn;
  final String summaryRu;
  final String outcomeEn;
  final String outcomeRu;
}

const List<CaseTypeDefinition> caseTypeRegistry = <CaseTypeDefinition>[
  CaseTypeDefinition(
    id: CaseTypeId.generalAdvisory,
    workflowMode: 'hybrid',
    views: <StudioCaseViewId>[
      StudioCaseViewId.issueMap,
      StudioCaseViewId.evidenceMap,
      StudioCaseViewId.decisionTable,
      StudioCaseViewId.timeline,
    ],
    requiredReview: 'professional',
    labelEn: 'Advisory decision',
    labelRu: 'Консультационное решение',
    summaryEn:
        'Structure issues, evidence, options and a reasoned recommendation.',
    summaryRu:
        'Структурируйте вопросы, доказательства, варианты и обоснованную рекомендацию.',
    outcomeEn: 'Decision memorandum',
    outcomeRu: 'Меморандум по решению',
  ),
  CaseTypeDefinition(
    id: CaseTypeId.litigationStrategy,
    workflowMode: 'adaptive',
    views: <StudioCaseViewId>[
      StudioCaseViewId.evidenceMap,
      StudioCaseViewId.timeline,
      StudioCaseViewId.decisionTable,
      StudioCaseViewId.taskPlan
    ],
    requiredReview: 'evidence_governance',
    labelEn: 'Litigation strategy',
    labelRu: 'Стратегия спора',
    summaryEn:
        'Connect claims, chronology, evidence, procedure, scenarios and deadlines.',
    summaryRu:
        'Свяжите требования, хронологию, доказательства, процедуру, сценарии и сроки.',
    outcomeEn: 'Strategy and risk report',
    outcomeRu: 'Отчёт о стратегии и рисках',
  ),
  CaseTypeDefinition(
    id: CaseTypeId.contractReview,
    workflowMode: 'decision',
    views: <StudioCaseViewId>[
      StudioCaseViewId.decisionTable,
      StudioCaseViewId.evidenceMap,
      StudioCaseViewId.taskPlan,
      StudioCaseViewId.timeline
    ],
    requiredReview: 'professional',
    labelEn: 'Contract review',
    labelRu: 'Проверка договора',
    summaryEn:
        'Map clauses, obligations, deviations, negotiation positions and residual risk.',
    summaryRu:
        'Сопоставьте условия, обязательства, отклонения, переговорные позиции и остаточный риск.',
    outcomeEn: 'Redline and risk report',
    outcomeRu: 'Отчёт по правкам и рискам',
  ),
  CaseTypeDefinition(
    id: CaseTypeId.taxPlanning,
    workflowMode: 'hybrid',
    views: <StudioCaseViewId>[
      StudioCaseViewId.issueMap,
      StudioCaseViewId.decisionTable,
      StudioCaseViewId.economics,
      StudioCaseViewId.timeline
    ],
    requiredReview: 'tax_governance',
    labelEn: 'Tax planning',
    labelRu: 'Налоговое планирование',
    summaryEn:
        'Compare lawful structures, current authorities, economics, substance and reporting obligations.',
    summaryRu:
        'Сравните законные структуры, актуальные источники, экономику, substance и отчётность.',
    outcomeEn: 'Tax position memorandum',
    outcomeRu: 'Меморандум по налоговой позиции',
  ),
  CaseTypeDefinition(
    id: CaseTypeId.compliance,
    workflowMode: 'process',
    views: <StudioCaseViewId>[
      StudioCaseViewId.taskPlan,
      StudioCaseViewId.decisionTable,
      StudioCaseViewId.evidenceMap,
      StudioCaseViewId.timeline
    ],
    requiredReview: 'evidence_governance',
    labelEn: 'Compliance assessment',
    labelRu: 'Compliance-оценка',
    summaryEn:
        'Map obligations to controls, evidence, exceptions, owners and remediation.',
    summaryRu:
        'Сопоставьте обязанности с контролями, доказательствами, исключениями, ответственными и исправлениями.',
    outcomeEn: 'Gap and remediation plan',
    outcomeRu: 'План устранения пробелов',
  ),
  CaseTypeDefinition(
    id: CaseTypeId.taxCompliance,
    workflowMode: 'hybrid',
    views: <StudioCaseViewId>[
      StudioCaseViewId.issueMap,
      StudioCaseViewId.decisionTable,
      StudioCaseViewId.economics,
      StudioCaseViewId.timeline,
    ],
    requiredReview: 'tax_governance',
    labelEn: 'Tax & compliance',
    labelRu: 'Налоги и compliance',
    summaryEn:
        'Compare lawful positions, economics, sources and reporting obligations.',
    summaryRu:
        'Сравните законные позиции, экономику, источники и обязанности по отчётности.',
    outcomeEn: 'Tax position memorandum',
    outcomeRu: 'Меморандум по налоговой позиции',
  ),
  CaseTypeDefinition(
    id: CaseTypeId.erpIncident,
    workflowMode: 'process',
    views: <StudioCaseViewId>[
      StudioCaseViewId.taskPlan,
      StudioCaseViewId.evidenceMap,
      StudioCaseViewId.decisionTable,
      StudioCaseViewId.timeline,
    ],
    requiredReview: 'professional',
    labelEn: 'ERP incident & solution',
    labelRu: 'ERP-инцидент и решение',
    summaryEn:
        'Capture the process failure, root cause, controls, solution and test evidence.',
    summaryRu:
        'Зафиксируйте сбой процесса, первопричину, контроли, решение и тестовые доказательства.',
    outcomeEn: 'Solution design & test pack',
    outcomeRu: 'Проект решения и пакет тестов',
  ),
  CaseTypeDefinition(
    id: CaseTypeId.investigation,
    workflowMode: 'adaptive',
    views: <StudioCaseViewId>[
      StudioCaseViewId.evidenceMap,
      StudioCaseViewId.timeline,
      StudioCaseViewId.taskPlan,
      StudioCaseViewId.decisionTable
    ],
    requiredReview: 'evidence_governance',
    labelEn: 'Investigation',
    labelRu: 'Расследование',
    summaryEn:
        'Preserve evidence, actors, chronology, contradictions, findings and unresolved matters.',
    summaryRu:
        'Сохраните доказательства, участников, хронологию, противоречия, выводы и открытые вопросы.',
    outcomeEn: 'Findings and chronology report',
    outcomeRu: 'Отчёт о выводах и хронологии',
  ),
  CaseTypeDefinition(
    id: CaseTypeId.trainingSimulation,
    workflowMode: 'simulation',
    views: <StudioCaseViewId>[
      StudioCaseViewId.simulation,
      StudioCaseViewId.timeline,
      StudioCaseViewId.evidenceMap,
    ],
    requiredReview: 'runtime_parity',
    labelEn: 'Training simulation',
    labelRu: 'Учебная симуляция',
    summaryEn:
        'Build a deterministic branching route with decisions, pressure and outcomes.',
    summaryRu:
        'Создайте детерминированный ветвящийся маршрут с решениями, давлением и исходами.',
    outcomeEn: 'Playable scenario',
    outcomeRu: 'Игровой сценарий',
  ),
];

CaseTypeDefinition caseTypeDefinition(CaseTypeId id) =>
    caseTypeRegistry.firstWhere((CaseTypeDefinition item) => item.id == id);

List<Map<String, dynamic>> caseTypeRegistrySignature() => caseTypeRegistry
    .map((CaseTypeDefinition item) => <String, dynamic>{
          'id': item.id.wireName,
          'version': caseTypeVersion,
          'workflowMode': item.workflowMode,
          'views': item.views
              .map((StudioCaseViewId view) => view.wireName)
              .toList(growable: false),
          'requiredReview': item.requiredReview,
        })
    .toList(growable: false);

File: app/case-type-registry.ts
import type { CaseTypeId, CaseTypeReference, CaseWorkflowMode, StudioDraft } from "./types";
import rawRegistry from "./case-type-registry.v1.json";
import { defaultTaxEconomics } from "./tax-economics";
import { caseTypeReference, CASE_TYPE_VERSION, DEFAULT_CASE_TYPE } from "./case-type-reference";
export { caseTypeReference, CASE_TYPE_REGISTRY_ID, CASE_TYPE_REGISTRY_SCHEMA_VERSION, CASE_TYPE_VERSION, DEFAULT_CASE_TYPE, normalizeCaseTypeReference } from "./case-type-reference";

export type CaseViewId = "issue_map" | "evidence_map" | "decision_table" | "task_plan" | "timeline" | "economics" | "simulation";

export type CaseTypeDefinition = {
  id: CaseTypeId;
  version: typeof CASE_TYPE_VERSION;
  workflowMode: CaseWorkflowMode;
  practiceArea: string;
  domain: "general" | "tax";
  views: readonly CaseViewId[];
  requiredReview: "professional" | "evidence_governance" | "tax_governance" | "runtime_parity";
};

const presentation: Record<CaseTypeId, { practiceArea: string; domain: "general" | "tax" }> = {
  general_advisory: { practiceArea: "General advisory", domain: "general" },
  litigation_strategy: { practiceArea: "Litigation strategy", domain: "general" },
  contract_review: { practiceArea: "Contract review", domain: "general" },
  tax_planning: { practiceArea: "Tax planning", domain: "tax" },
  compliance: { practiceArea: "Compliance", domain: "general" },
  tax_compliance: { practiceArea: "International tax planning", domain: "tax" },
  erp_incident: { practiceArea: "ERP incident & solution design", domain: "general" },
  investigation: { practiceArea: "Investigation", domain: "general" },
  training_simulation: { practiceArea: "Professional training", domain: "general" },
};

type RawRegistry = { format: string; schemaVersion: number; registry: string; types: Array<Omit<CaseTypeDefinition, "practiceArea" | "domain">> };

function validatedRegistry(value: unknown): readonly CaseTypeDefinition[] {
  const registry = value as RawRegistry;
  if (!registry || registry.format !== "genesis-juris-case-type-registry" || registry.schemaVersion !== 1 || registry.registry !== "genesis-juris-case-types" || !Array.isArray(registry.types)) throw new Error("Unsupported case-type registry");
  const ids = new Set<CaseTypeId>();
  return registry.types.map((entry) => {
    if (!(entry.id in presentation) || entry.version !== CASE_TYPE_VERSION || ids.has(entry.id)) throw new Error("Invalid immutable case-type package");
    ids.add(entry.id);
    return { ...entry, ...presentation[entry.id] };
  });
}

export const CASE_TYPE_REGISTRY = validatedRegistry(rawRegistry);

const byId = new Map(CASE_TYPE_REGISTRY.map((definition) => [definition.id, definition]));

export function caseTypeDefinition(reference?: CaseTypeReference | null) {
  return byId.get(reference?.id ?? DEFAULT_CASE_TYPE.id) ?? byId.get(DEFAULT_CASE_TYPE.id)!;
}

export function applyCaseType(draft: StudioDraft, id: CaseTypeId): StudioDraft {
  const definition = byId.get(id);
  if (!definition) throw new Error("Unknown case type");
  const classification = draft.classification ?? {
    practiceArea: definition.practiceArea,
    difficulty: "Intermediate",
    tags: [],
    taxTopics: [],
    complianceOnly: true,
  };
  const nextClassification = {
    ...classification,
    domain: definition.domain,
    practiceArea: definition.practiceArea,
    complianceOnly: true,
    ...(definition.domain === "tax"
      ? { purpose: classification.purpose ?? "compliance_review" as const }
      : { taxTopics: [], purpose: "compliance_review" as const }),
  };
  return {
    ...draft,
    caseType: caseTypeReference(id),
    classification: nextClassification,
    ...(definition.domain === "tax"
      ? { taxEconomics: draft.taxEconomics ?? defaultTaxEconomics(draft.dealEconomics?.currency ?? "EUR") }
      : { taxEconomics: undefined }),
  };
}

export function caseTypeRegistrySignature() {
  return CASE_TYPE_REGISTRY.map(({ id, version, workflowMode, views, requiredReview }) => ({ id, version, workflowMode, views: [...views], requiredReview }));
}
